File: app/android/app/src/main/kotlin/com/missingbulb/setlist/CaptureRecorder.kt
package com.missingbulb.setlist

import android.content.Context
import android.media.MediaRecorder
import android.os.Build
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File
import kotlin.math.max
import kotlin.math.min

/**
 * The Android half of the recording seam.
 *
 * `MediaRecorder` rather than `AudioRecord` (decision D10): nothing analyses
 * while capture runs, so no part of the app needs raw PCM, and the high-level
 * recorder already writes AAC and reports an amplitude. Audio never crosses
 * into Dart — only the path to write, the seconds captured, a level, and the
 * interruption transitions.
 *
 * Unlike iOS there is no interruption callback to subscribe to: under the
 * shared-microphone policy a privileged app may take the input while this
 * recorder keeps running and records silence (risk R9). What that means for
 * requirement 6.6 is unresolved, and is the first thing to settle on a device.
 */
class CaptureRecorder : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private var appContext: Context? = null
    private var recorder: MediaRecorder? = null
    private var events: EventChannel.EventSink? = null
    private val handler = Handler(Looper.getMainLooper())
    private var startedAtMillis = 0L
    private var capturedBeforeThisPart = 0.0

    private val meter = object : Runnable {
        override fun run() {
            val active = recorder ?: return
            events?.success(
                mapOf(
                    "type" to "level",
                    // getMaxAmplitude is 0..32767 and resets on each read, so
                    // this is the loudest sample since the previous tick.
                    "level" to min(1.0, max(0.0, active.maxAmplitude / 32767.0)),
                    "captured" to captured(),
                ),
            )
            handler.postDelayed(this, METER_INTERVAL_MILLIS)
        }
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        appContext = binding.applicationContext
        MethodChannel(binding.binaryMessenger, METHOD_CHANNEL).setMethodCallHandler(this)
        EventChannel(binding.binaryMessenger, EVENT_CHANNEL).setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        stop()
        appContext = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "start" -> {
                val path = call.argument<String>("path")
                if (path == null) {
                    result.error("no-path", "start needs a path", null)
                    return
                }
                runCatching { start(path) }
                    .onSuccess { result.success(null) }
                    .onFailure { result.error("start-failed", it.message, null) }
            }
            "stop" -> {
                stop()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, sink: EventChannel.EventSink?) {
        events = sink
    }

    override fun onCancel(arguments: Any?) {
        events = null
    }

    private fun start(path: String) {
        val file = File(path)
        file.parentFile?.mkdirs()

        val context = appContext
        val active = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && context != null) {
            MediaRecorder(context)
        } else {
            @Suppress("DEPRECATION")
            MediaRecorder()
        }
        active.setAudioSource(MediaRecorder.AudioSource.MIC)
        active.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
        active.setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
        active.setAudioChannels(1)
        active.setAudioSamplingRate(44_100)
        active.setAudioEncodingBitRate(128_000)
        active.setOutputFile(path)
        active.prepare()
        active.start()

        recorder = active
        startedAtMillis = System.currentTimeMillis()
        handler.post(meter)
    }

    private fun stop() {
        handler.removeCallbacks(meter)
        recorder?.let { active ->
            runCatching { active.stop() }
            active.release()
            // The clock counts recorded audio across every part of the show, so
            // what this part captured is carried forward rather than reset.
            capturedBeforeThisPart = captured()
        }
        recorder = null
    }

    private fun captured(): Double =
        if (recorder == null) {
            capturedBeforeThisPart
        } else {
            capturedBeforeThisPart + (System.currentTimeMillis() - startedAtMillis) / 1000.0
        }

    companion object {
        const val METHOD_CHANNEL = "com.missingbulb.setlist/capture"
        const val EVENT_CHANNEL = "com.missingbulb.setlist/capture-events"
        private const val METER_INTERVAL_MILLIS = 100L
    }
}

File: app/ios/Runner/CaptureRecorder.swift
import AVFoundation
import Flutter

/// The iOS half of the recording seam.
///
/// `AVAudioRecorder` rather than `AVAudioEngine` (decision D10): nothing
/// analyses while capture runs, so no part of the app needs raw PCM, and the
/// high-level recorder already writes AAC and reports a level. Audio never
/// crosses into Dart — only the path to write, the seconds captured, a level,
/// and the interruption transitions.
final class CaptureRecorder: NSObject {
    static let methodChannel = "com.missingbulb.setlist/capture"
    static let eventChannel = "com.missingbulb.setlist/capture-events"

    private var recorder: AVAudioRecorder?
    private var meterTimer: Timer?
    private var sink: FlutterEventSink?

    func register(with registrar: FlutterPluginRegistrar) {
        FlutterMethodChannel(name: Self.methodChannel, binaryMessenger: registrar.messenger())
            .setMethodCallHandler { [weak self] call, result in
                self?.handle(call, result: result)
            }
        FlutterEventChannel(name: Self.eventChannel, binaryMessenger: registrar.messenger())
            .setStreamHandler(self)

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleInterruption(_:)),
            name: AVAudioSession.interruptionNotification,
            object: AVAudioSession.sharedInstance()
        )
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "start":
            guard let path = (call.arguments as? [String: Any])?["path"] as? String else {
                result(FlutterError(code: "no-path", message: "start needs a path", details: nil))
                return
            }
            do {
                try start(atPath: path)
                result(nil)
            } catch {
                result(FlutterError(code: "start-failed", message: "\(error)", details: nil))
            }
        case "stop":
            stop()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func start(atPath path: String) throws {
        let url = URL(fileURLWithPath: path)
        try FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )

        let session = AVAudioSession.sharedInstance()
        // `.record` with the background-audio entitlement is what keeps capture
        // alive behind a locked screen (requirement 6.5).
        try session.setCategory(.record, mode: .default)
        try session.setActive(true)

        let recorder = try AVAudioRecorder(url: url, settings: [
            AVFormatIDKey: kAudioFormatMPEG4AAC,
            AVSampleRateKey: 44_100,
            AVNumberOfChannelsKey: 1,
            AVEncoderBitRateKey: 128_000,
        ])
        recorder.isMeteringEnabled = true
        recorder.record()
        self.recorder = recorder
        startMetering()
    }

    private func stop() {
        meterTimer?.invalidate()
        meterTimer = nil
        recorder?.stop()
        recorder = nil
        try? AVAudioSession.sharedInstance().setActive(false)
    }

    private func startMetering() {
        meterTimer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { [weak self] _ in
            guard let self, let recorder = self.recorder else { return }
            recorder.updateMeters()
            self.sink?([
                "type": "level",
                // averagePower is in decibels, -160...0. The strip wants a
                // 0...1 reading, and below -50 dB there is nothing to draw.
                "level": max(0, (recorder.averagePower(forChannel: 0) + 50) / 50),
                "captured": recorder.currentTime,
            ])
        }
    }

    @objc private func handleInterruption(_ notification: Notification) {
        guard let raw = notification.userInfo?[AVAudioSessionInterruptionTypeKey] as? UInt,
              let type = AVAudioSession.InterruptionType(rawValue: raw) else { return }
        switch type {
        case .began:
            meterTimer?.invalidate()
            sink?(["type": "interrupted", "cause": "microphoneTaken",
                   "captured": recorder?.currentTime ?? 0])
        case .ended:
            // The coordinator opens a new part rather than resuming into the
            // old file: the clock must not rewind across the gap.
            sink?(["type": "resumed", "captured": recorder?.currentTime ?? 0])
        @unknown default:
            break
        }
    }
}

extension CaptureRecorder: FlutterStreamHandler {
    func onListen(withArguments _: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
        sink = eventSink
        return nil
    }

    func onCancel(withArguments _: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
